### PR TITLE
Show package count in header

### PR DESCRIPTION
**Pull request by JakobDev**
This shows the number of packages in the header. This applies to repos, users and orgs. The count is also returned from the API.

grafik

I haven't found a way to connect a package to a repo through the API, so I couldn't write a test for that.

### CODE REVIEW

File: models/packages/package_version.go [hunk 1]
```diff
@@ -298,8 +299,7 @@ func SearchVersions(ctx context.Context, opts *PackageSearchOptions) ([]*Package
 	return pvs, count, err
 }
 
-// SearchLatestVersions gets the latest version of every package matching the search options
-func SearchLatestVersions(ctx context.Context, opts *PackageSearchOptions) ([]*PackageVersion, int64, error) {
+func getLatestVersions(ctx context.Context, opts *PackageSearchOptions) *xorm.Session {
```
Comment (KN4CK3R) [inline models/packages/package_version.go]: I'm not really pleased by this name. Maybe `buildLatestVersionsSession` or something similar?

File: templates/org/menu.tmpl [hunk 2]
```diff
@@ -14,6 +14,9 @@
 		{{if and .IsPackageEnabled .CanReadPackages}}
 		<a class="item" href="{{$.Org.HomeLink}}/-/packages">
 			{{svg "octicon-package"}} {{.locale.Tr "packages.title"}}
+			{{if .PackageCount}}
```
Comment (yp05327) [inline templates/org/menu.tmpl:17]: `NumXXXX` `XXXXCount` which is better?
(not a block, just a question)

Comment (JakobDev) [reply templates/org/menu.tmpl:17]: I think they are both equal

File: modules/context/repo.go [hunk 3]
```diff
@@ -17,6 +17,7 @@ import (
 	"code.gitea.io/gitea/models/db"
 	git_model "code.gitea.io/gitea/models/git"
 	issues_model "code.gitea.io/gitea/models/issues"
+	packages_model "code.gitea.io/gitea/models/packages"
 	access_model "code.gitea.io/gitea/models/perm/access"
 	repo_model "code.gitea.io/gitea/models/repo"
```
Comment (delvh) [inline modules/context/repo.go:22]: This file needs a **serious** refactoring.
A `modules/context` file, which is one of the most essential files in the hierarchy, should not depend on `models` and `services`…
But that's not a problem for this PR.

File: services/convert/convert.go [hunk 4]
```diff
@@ -283,6 +284,8 @@ func ToDeployKey(apiLink string, key *asymkey_model.DeployKey) *api.DeployKey {
 
 // ToOrganization convert user_model.User to api.Organization
 func ToOrganization(ctx context.Context, org *organization.Organization) *api.Organization {
+	numPackages, _ := packages_model.CountLatestVersions(ctx, &packages_model.PackageSearchOptions{OwnerID: org.ID, IsInternal: util.OptionalBoolFalse})
```
Comment (delvh) [inline services/convert/convert.go:287]: This change looks really strange and not at all how the code should be structured.
This method **should not** make database queries.
It's a much better idea to store the information inside `organization.Organization`, **if at all**:
After all, there are already API routes for getting this information (-> `X-Total-Count`).
Furthermore, if we do that, I can already see the requests to also add it for issues, labels, milestones, PRs, releases, tags, projects, …
So in summary, I'm against modifying the API.

File: services/convert/repository.go [hunk 5]
```diff
@@ -193,6 +201,7 @@ func innerToRepo(ctx context.Context, repo *repo_model.Repository, permissionInR
 		OpenIssues:                    repo.NumOpenIssues,
 		OpenPulls:                     repo.NumOpenPulls,
 		Releases:                      int(numReleases),
+		Packages:                      int(numPackages),
```
Comment (delvh) [inline services/convert/repository.go:204]: See comment above.

File: services/convert/user.go [hunk 6]
```diff
@@ -62,6 +66,7 @@ func toUser(ctx context.Context, user *user_model.User, signed, authed bool) *ap
 		Followers:    user.NumFollowers,
 		Following:    user.NumFollowing,
 		StarredRepos: user.NumStars,
+		Packages:     int(numPackages),
```
Comment (delvh) [inline services/convert/user.go:69]: See comment above.

**Review by delvh — CHANGES_REQUESTED** [review]

File: services/convert/convert.go (hunk 4 above)
Comment (JakobDev) [reply services/convert/convert.go:287]: To the API response in general: The Number is shown on the Profile, so it should also be returned by the API, so 3rd Party Apps can add this Feature easily. As we don't show Issues or Milestones on the profile, there is no need to return a count.

To the location: There are also other places in the convert API ehere databse querrys are done, but I can move it to the User7org Modell and add a `LoadCount` function.

**Comment by wxiaoguang** [comment]
I'd like to say that we shouldn't do so.

Actually we should remove most useless from the tab titles, also reduce unnecessary database queries.

* Useful numbers: issue, PR, security
* Useless numbers: package, release
    * GitHub doesn't show those numbers either
    * End users do not really care about how many "releases" there are, most people only need the "latest" one.

And this PR is stale for long time, so I think we can close it. Feel free to reopen if there are better designs.
